File: tutorials/elixir/04-hello-tui/lib/hello_tui.ex
defmodule HelloTui do
  @moduledoc """
  A minimal terminal UI app built with Owl.
  Displays a selectable list with arrow-key navigation.
  """

  alias HelloTui.Model

  def start do
    model = Model.new()
    loop(model)
  end

  defp loop(model) do
    render(model)

    case IO.getn("", 1) do
      "q" ->
        IO.puts("\nGoodbye!")

      "\e" ->
        # Read escape sequence
        case IO.getn("", 2) do
          "[A" -> loop(Model.move_up(model))
          "[B" -> loop(Model.move_down(model))
          _ -> loop(model)
        end

      _ ->
        loop(model)
    end
  end

  defp render(%Model{items: items, cursor: cursor} = model) do
    # Clear screen and move to top
    IO.write("\e[2J\e[H")
    IO.puts(Owl.Data.tag("Hello TUI - Use arrows to navigate, q to quit", :cyan))
    IO.puts(String.duplicate("─", 50))

    items
    |> Enum.with_index()
    |> Enum.each(fn {item, idx} ->
      if idx == cursor do
        IO.puts(Owl.Data.tag("▸ #{item}", [:green, :bold]))
      else
        IO.puts("  #{item}")
      end
    end)

    IO.puts(String.duplicate("─", 50))
    IO.puts("Selected: #{Owl.Data.tag(Model.selected_item(model), :yellow)}")
  end
end

File: tutorials/erlang/04-hello-tui/src/hello_tui.erl
-module(hello_tui).
-export([start/0]).

start() ->
    io:setopts([{binary, true}]),
    State = hello_tui_state:new(),
    loop(State).

loop(State) ->
    Output = hello_tui_render:render(State),
    io:put_chars(Output),
    case read_key() of
        quit ->
            io:put_chars("\e[2J\e[H"),
            io:format("You selected: ~s~n", [hello_tui_state:selected_item(State)]),
            ok;
        up ->
            loop(hello_tui_state:move_up(State));
        down ->
            loop(hello_tui_state:move_down(State));
        _ ->
            loop(State)
    end.

read_key() ->
    case io:get_chars(<<>>, 1) of
        <<"q">> -> quit;
        <<"\e">> ->
            case io:get_chars(<<>>, 1) of
                <<"[">> ->
                    case io:get_chars(<<>>, 1) of
                        <<"A">> -> up;
                        <<"B">> -> down;
                        _ -> unknown
                    end;
                _ -> unknown
            end;
        _ -> unknown
    end.
